plutus-tx/src/PlutusTx/AssocMap.hs

```diff
@@ -55,195 +51,241 @@ import Prettyprinter (Pretty (..))
-          traverseFromBuiltin :: BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> Maybe [(k, v)]
-          traverseFromBuiltin = go
-            where
-                go :: BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> Maybe [(k, v)]
-                go l = BI.chooseList l (const (pure [])) (\_ -> let tup = BI.head l in liftA2 (:) (liftA2 (,) (fromBuiltinData $ BI.fst tup) (fromBuiltinData $ BI.snd tup)) (go (BI.tail l))) ()
+  fromBuiltinData d =
+    P.matchData'
+      d
+      (\_ _ -> Nothing)
+      (\es -> Map <$> traverseFromBuiltin es)
+      (const Nothing)
+      (const Nothing)
+      (const Nothing)
+    where
+      {-# INLINE traverseFromBuiltin #-}
+      traverseFromBuiltin ::
+        BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) ->
+        Maybe [(k, v)]
+      traverseFromBuiltin = go
+        where
+          go :: BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> Maybe [(k, v)]
+          go l =
+            BI.chooseList
+              l
+              (const (pure []))
+              ( \_ ->
+                  let ~tup = BI.head l
+                   in liftA2
+                        (:)
+                        (liftA2 (,) (fromBuiltinData $ BI.fst tup) (fromBuiltinData $ BI.snd tup))
+                        (go (BI.tail l))
+              )
+              ()
 
 instance (UnsafeFromData k, UnsafeFromData v) => UnsafeFromData (Map k v) where
-    unsafeFromBuiltinData d = let es = BI.unsafeDataAsMap d in Map $ mapFromBuiltin es
-      where
-          {-# INLINE mapFromBuiltin #-}
-          mapFromBuiltin :: BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> [(k, v)]
-          mapFromBuiltin = go
-            where
-                go ::  BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> [(k, v)]
-                go l = BI.chooseList l (const []) (\_ -> let tup = BI.head l in (unsafeFromBuiltinData $ BI.fst tup, unsafeFromBuiltinData $ BI.snd tup) : go (BI.tail l)) ()
+  -- The `~` here enables `BI.unsafeDataAsMap d` to be inlined, which reduces costs slightly.
+  -- Without the `~`, the inliner would consider it not effect safe to inline.
+  -- The other `~`s in this module are there for the same reason.
+  -- We can remove all the `~`s once we make the inliner smart enough to inline them.
+  -- See https://github.com/input-output-hk/plutus/pull/5371#discussion_r1297833685
+  unsafeFromBuiltinData d = let ~es = BI.unsafeDataAsMap d in Map $ mapFromBuiltin es
+    where
+      {-# INLINE mapFromBuiltin #-}
+      mapFromBuiltin :: BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> [(k, v)]
+      mapFromBuiltin = go
+        where
+          go :: BI.BuiltinList (BI.BuiltinPair BI.BuiltinData BI.BuiltinData) -> [(k, v)]
+          go l =
+            BI.chooseList
+              l
+              (const [])
+              ( \_ ->
+                  let ~tup = BI.head l
```

Comment:
Yeah, good question. Without the `~` here, we get some very weird Core for `unsafeFromData @(Map k v)`. It basically looks like this:

```
-- GHC Core
...
let lvl9 = BI.head l
 in ...
     BI.chooseList
     l
     (const [])
     (\_ -> case lvl9 of _ -> ...fst lvl9...snd lvl9...)
```

which means `lvl9` is evaluated strictly, but then the result is not used. Instead, `lvl9` is re-evaluated at `fst` and `snd`!

I don't know why, but I'll add a note here.

Reply:
I'm a bit confused: it doesn't look like `lvl9` is evaluated strictly there. It looks like we've got the classic strict-binding creation of:
```
let ~lvl9ns = BI.head l
     !lvl9s = lvl9
in ... lvl9s ... lvl9s
```
and then we lower the scope of one of the bindings. So that looks fine to me?

Reply:
(I think maybe the confusion is that in Core, let-bindings are still non-strict. The effect of `Strict` is basically inserting `case` expressions to force evaluation of stuff, which is what we see later down, and which get turned into strict let-bindings by the plugin. Then the PIR compiler can inline the non-strict binding and we should get something sensible.)